### PR TITLE
Refactor DetailsRowFields to remove redundant columns iteration and setState call

### LINKED ISSUE #7664: Reduce DetailsList's DetailsRowFields render time and memory footprint
**Describe the feature that you would like added**

**Performance:** The `DetailsList`'s `DetailsRowFields` component renders each row's cell contents except for the "selection" checkbox. Currently, the component iterates over `N` columns _twice_ per render. It also unnecessarily allocates an array of size `N` which contains the elements to render per cell. It _also_ invokes `setState` within `componentWillReceiveProps` when it could handle all of the work within `render` via props.

**What component or utility would this be added to:** `DetailsList`, specifically `DetailsRowFields`

**Have you discussed this feature with our team, and if so, who** Yes, with `DetailsList` code-owners.

**Additional context/screenshots**

The highlighted sections of the below flame-graph collected via the Chrome profiler on scroll of a virtualized `DetailsList` shows the time spent due to redundant columns array iteration.

image

After removal, the time spent rendering `DetailsRowFields` is reduced, see below:

image

### PR DESCRIPTION
**Pull request by KevinTCoughlin**
#### Pull request checklist

- [x] Addresses an existing issue: Fixes #7664
- [x] Include a change request file using `$ npm run change`

#### Description of changes

`DetailsRowFields` is iterating twice over `N` columns per row render. This change-set does the following:

1. Removes `componentWillReceiveProps` usage in favor of in-line in `render`, removing `setState` call.
1. Removes _redundant_ iteration of columns performed in `_getState` and allocation.
1. Hoists the `getCellText` function so there isn't one _per_ `DetailsRowFields`.
1. Corrects the return value of `getCellText` from `void` to `string`
1. ~~Simplifies `getCellText` logic.~~ Edit: Reverted, need to handle `0` case.
1. Removes `IDetailsRowFieldsState` declaration since no longer used with refactor.
1. Removes the `try..catch` surrounding the cell render function

I'm most unsure about **item 7** above without additional context. cc: @aditima @dzearing 

#### Performance Profiler

Local dev URL used to test: http://localhost:4322/#/tests/detailslistbasicexample

**Before**

I decided to look at `DetailsRowFields` after running Chrome's Profiler when scrolling the virtualized `DetailsList`. I noticed that `DetailsRowFields` was spending _some_ time in `componentWillReceiveProps` (highlighted below) which surprised me given it seemed like a simple pass-thru for _most cases_.

image

**After**

After this refactor, the same profiler snapshot no longer shows time spent in `componentWillReceiveProps`.

image

#### Focus areas to test

I'm operating under the assumption that all LoC exist for a reason, so I need help from those with more React knowledge and context. Please do let me know if this refactor risks regression.

I performed the below cursory tests _before_ and _after_ this change-set and compared behavior.

**Tests Ran**

1. Changing `item.onRender` for _ALL_ items on toggle interaction.
1. Render _odd_ rows differently than _even_ rows via `item.onRender`.
1. Create new `items` array with a single item containing custom name.
1. Single selection with virtualization. Scrolled fully `DOWN` and `UP` the list. Selection was retained. 
1. Multiple selection with virtualization. Scolled fully `DOWN` and `UP` the list. Selection was retained. 

The above "tests" exhibited the same behavior with and without this change-set.

### CODE REVIEW

**Comment by KevinTCoughlin** [comment]
Ah, need to revert https://github.com/OfficeDev/office-ui-fabric-react/pull/7649/commits/fa9db0e1f8d71245507a25fe01b7d49159471c71 for the case of `0`. Good catch tests :).

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx [hunk 1]
```diff
@@ -53,41 +59,11 @@ export class DetailsRowFields extends BaseComponent<IDetailsRowFieldsProps, IDet
...
-
-    return {
-      cellContent: props.columns.map(column => {
-        let cellContent;
-
-        try {
-          const render = column.onRender || onRenderItemColumn;
-
-          cellContent = render && !shimmer ? render(item, itemIndex, column) : this._getCellText(item, column);
-        } catch (e) {
```
Comment (Vitalius1) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx]: Wasn't the try/catch here for cases when neither `column.onRender` or `onRenderItemColumn` are provided? If it's `undefined` you might get an error `undefined is not a function` in this edge case.

Comment (Vitalius1) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx]: From the history, it looks like it's there from the very beginning when David created it.

**Comment by KevinTCoughlin** [comment]
@ThomasMichon shared some additional ideas for improving performance here which I'd like to capture:

1. Performance gain in having `this.props.onRenderItemColumn` take precedence over `column.onRender`.
1. When possible (perhaps here) extend `React.PureComponent` instead of `BaseComponent`.

I have a commit which I removed from this branch which refactors `DetailsRowFieldRows` to extend from `PureComponent`. That may be worth exploring in a separate iteration. I'm unsure about the first item re: backwards compatibility, need to investigate further.

cc: @dzearing @aditima

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx (hunk 1 above)
Comment (KevinTCoughlin) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx]: Indeed, that's why I'm looking for if anyone has context. It seems like we'd only have an issue if the provided render function throws. Not sure if that's something worth guarding against.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx [hunk 2]
```diff
@@ -35,6 +38,9 @@ export class DetailsRowFields extends BaseComponent<IDetailsRowFieldsProps, IDet
                 cellStyleProps.cellRightPadding +
                 (column.isPadded ? cellStyleProps.cellExtraRightPadding : 0);
 
+          const { onRender = onRenderItemColumn } = column;
+          const cellContentsRender = onRender && !shimmer ? onRender(item, itemIndex, column) : getCellText(item, column);
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx]: Can you error on verbosity and add parenthesis?

(onRender && !shimmer) ? ... : ...

or

onRender && (!shimmer ? ... : ...)

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx (hunk 1 above)
Comment (dzearing) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx]: This was written in the time of react 0.14.

Comment (dzearing) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx]: I think it's ok to remove.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx (hunk 2 above)
Comment (kenotron) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx]: I THIHK prettier strips things out ...

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx [hunk 3]
```diff
@@ -4,25 +4,28 @@ import { BaseComponent, css } from '../../Utilities';
...
 
-export interface IDetailsRowFieldsState {
-  cellContent: React.ReactNode[];
-}
-
-export class DetailsRowFields extends BaseComponent<IDetailsRowFieldsProps, IDetailsRowFieldsState> {
-  constructor(props: IDetailsRowFieldsProps) {
-    super(props);
+const getCellText = (item: any, column: IColumn): string => {
+  let value = item && column && column.fieldName ? item[column.fieldName] : '';
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx]: maybe parenthesis

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx (hunk 2 above)
Comment (KevinTCoughlin) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx]: > 
> 
> I THIHK prettier strips things out ...

prettier does remove the parens from the first case on commit

```tsx
const cellContentsRender = (onRender && !shimmer) ? onRender(item, itemIndex, column) : getCellText(item, column);
```

The second suggestion seems to clash with the intent of the ternary, but perhaps I'm misunderstanding.

```tsx
const cellContentsRender = onRender && (!shimmer ? onRender(item, itemIndex, column) : getCellText(item, column));
```

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx (hunk 1 above)
Comment (KevinTCoughlin) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx]: Thank you for providing context :)

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx (hunk 2 above)
Comment (KevinTCoughlin) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx]: Spoke with David offline, sticking with what we have. Resolving.

**Pull request merged**